Python/ast_opt.c

```diff
@@ -607,12 +607,21 @@ fold_iter(expr_ty arg, PyArena *arena, _PyASTOptimizeState *state)
         arg->v.Tuple.ctx = ctx;
         /* Try to create a constant tuple. */
         newval = make_const_tuple(elts);
+        if (newval == NULL) {
+            return 0;
+        }
```

Comment:
In `fold_tuple`, this call is immediately followed by `make_const` which then clears any exception that is *not* PyExc_KeyboardInterrupt, and then returns 1. So my question is: should we actually clear the error or not in this case as well? (or maybe patch `fold_tuple` in this case?).

In other words, calls to `make_const` are not preceeded by `newval == NULL` checks.

Reply:
The fact that `make_const` does error checking for one of their arguments and callers are relying on that behaviour is not only error prone but it's also very confusing. If we fail to create one of this objects we should propagate the error (think about MemoryError or something worse, we **surely** don't want to clean that). 

I don't understand also why we are just special-casing KeyboardInterrupt here and not in other parts.

> In other words, calls to make_const are not preceeded by newval == NULL checks.

I think that's fundamentally flawed. This code was moved from the time this happened in `peephole.c` (and even before, if you track it down it goes as back as 00ee7baf4943) and was clearing out all exceptions in the peephole after compilation and special casing `KeyboardInterrupt` which it's masking real errors and can actually crash if the caller is left in an inconsistent state.

Reply:
We should clean this up in another PR

Reply:
Actually, I just wanted to address the test failures (which I *think* they could be related to this). But I agree that we shouldn't clear errors otherwise. I was more concerned about: "should we fix calls to `make_const` here or separately?".

Reply:
We should always propagate. If there is an exception we screwed up somewhere and the code should defend against it otherwise. Reliying on a catch-all it's bad because this is basically ignoring anything inheriting from BaseException, including SystemErrors

Reply:
Here the other problem really is that `make_const_tuple` it's returning `NULL` without the error set :(

Reply:
I have made here a big refactor so `make_const` doesn't eat errors and so exceptions that are explicitly ignored are always ignored